### PR TITLE
feat: add combination test infrastructure and basic tests

### PR DESCRIPTION
**Pull request by lovywu**
## Summary

- Add multi-operator combination testing framework for FlagGems, including shared infrastructure (`accuracy_utils`, `conftest`, `logging_config`, `utils/numerical_stability`)
- Add 6 basic test suites covering: attention numerical stability, convolution combination, loss functions, mixed precision accumulation, sampling operations, and sequence operations
- These tests validate operator behavior when multiple operators are chained together (e.g., softmax → attention → masking, conv → BN → ReLU)

## Files

**Infrastructure (6 files):**
- `tests/combination/__init__.py`
- `tests/combination/conftest.py` — shared fixtures (`use_gems`, `dtype`, `batch_size`, etc.)
- `tests/combination/accuracy_utils.py` — reference computation (GPU fp64 / CPU fallback) and tolerance scaling
- `tests/combination/logging_config.py` — structured JSON test logger
- `tests/combination/utils/__init__.py`
- `tests/combination/utils/numerical_stability.py` — NaN/Inf/gradient health checks

**Test suites (6 files):**
- `test_attention_numerical_stability.py` — softmax/attention edge cases (padding mask, causal mask, precision)
- `test_convolution_combination.py` — conv2d/conv3d + pooling chains, residual blocks
- `test_loss_functions.py` — cross-entropy, MSE, contrastive loss, multi-task gradient flow
- `test_mixed_precision_accumulation.py` — fp16/bf16 error accumulation bounds
- `test_sampling_operations.py` — dropout, weight initialization, multinomial sampling
- `test_sequence_operations.py` — embedding, gather, cumulative ops (cumsum/cummax/cummin)

## Note

This is the first of a series of PRs for the combination testing framework. Subsequent PRs will add:
- MoE routing tests (depends on this PR)
- Model definitions + model-dependent tests (depends on this PR)
- CI workflow and documentation
- Operator extraction tools

## Test plan

- [ ] `pytest tests/combination/test_loss_functions.py -v`
- [ ] `pytest tests/combination/test_sampling_operations.py -v`
- [ ] `pytest tests/combination/test_sequence_operations.py -v`
- [ ] `pytest tests/combination/test_convolution_combination.py -v`
- [ ] `pytest tests/combination/test_mixed_precision_accumulation.py -v`
- [ ] `pytest tests/combination/test_attention_numerical_stability.py -v`

### CODE REVIEW

File: tests/combination/accuracy_utils.py [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+import logging
+import math
+
+import torch
+
+import flag_gems
+from flag_gems.testing import RESOLUTION
+
+fp64_is_supported = flag_gems.runtime.device.support_fp64
+bf16_is_supported = flag_gems.runtime.device.support_bf16
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:26]: are we sure that these `support_fp64` and `support_bf16` attributes are accessible from `flag_gems.runtime.device`?

better do a `try...except` here

File: tests/combination/accuracy_utils.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    COMBO_FLOAT_DTYPES.append(torch.bfloat16)
+
+# Low-precision only (for mixed-precision accumulation tests that compare against fp32).
+COMBO_LOW_PRECISION_DTYPES = [torch.float16]
+if bf16_is_supported:
+    COMBO_LOW_PRECISION_DTYPES.append(torch.bfloat16)
+
+try:
+    from ..conftest import TO_CPU
+except ImportError:
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:40]: Will this happen at all?

File: tests/combination/accuracy_utils.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+# Combination test float dtypes – adapts to backend capabilities.
+COMBO_FLOAT_DTYPES = [torch.float16, torch.float32]
+if bf16_is_supported:
+    COMBO_FLOAT_DTYPES.append(torch.bfloat16)
+
+# Low-precision only (for mixed-precision accumulation tests that compare against fp32).
+COMBO_LOW_PRECISION_DTYPES = [torch.float16]
+if bf16_is_supported:
+    COMBO_LOW_PRECISION_DTYPES.append(torch.bfloat16)
+
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:37]: What are we gonna do with `float64`、`complex32`、`complex64`?

File: tests/combination/accuracy_utils.py (hunk 1 above)
Comment (tengqm) [reply tests/combination/accuracy_utils.py:26]: We are not considering the `fp64_is_supported` later on?

File: tests/combination/accuracy_utils.py [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+                o.detach().to(orig_device) if isinstance(o, torch.Tensor) else o
+                for o in ref_output
+            )
+
+        return ref_output, input_grads, param_grads
+
+    try:
+        return _run(ref_device, ref_dtype)
+    except RuntimeError:
+        return _run("cpu", torch.float32)
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:233]: I can understand that this fallback is intentional ... but I'm wondering if a silent fallback is appropriate.

File: tests/combination/accuracy_utils.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+            extra={
+                "test_data": {
+                    "event": "accuracy_check",
+                    "check_type": "forward_comparison",
+                    "name": name,
+                    "dtype": str(dtype),
+                    "num_ops": num_ops,
+                    "expected_atol": atol_val,
+                    "expected_rtol": rtol_val,
+                    "actual_max_error": max_err,
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:284]:
```suggestion
                    "max_error": max_err,
```

File: tests/combination/accuracy_utils.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+                    "actual_mean_error": mean_err,
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:285]: For all field names, the shorter, the better. This is especially true for fields that are in a well-defined context.

```suggestion
                    "mean_error": mean_err,
```

File: tests/combination/accuracy_utils.py [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+            ref,
+            atol=atol_val,
+            rtol=rtol_val,
+            msg=lambda m: f"{name}: {m}" if name else m,
+        )
+    except AssertionError:
+        passed = False
+        raise
+    finally:
+        logger.info(
```
Comment (tengqm) [inline tests/combination/accuracy_utils.py:273]: I'm wondering if we should use `logger` at all.
We are risking mixing the logs from out code with whatever other libraries/modules produce.
An alternative is to just specify a global *record* file, dump all your outputs to that file, in pretty, clean JSON for postmortem checking.

File: tests/combination/accuracy_utils.py (hunk 2 above)
Comment (tengqm) [reply tests/combination/accuracy_utils.py:40]: Later on you have your own copy of `TO_CPU` ...

File: tests/combination/conftest.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+# Fixtures for common test configurations
+# ============================================================================
+
+
+@pytest.fixture(scope="module")
+def use_gems():
+    """Context manager to enable FlagGems for a test module."""
+    with flag_gems.use_gems():
+        yield
+
```
Comment (tengqm) [inline tests/combination/conftest.py:94]: Is this safe?

File: tests/combination/conftest.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    )
+    config.addinivalue_line("markers", "ffn: mark test related to FFN modules")
+
+    # Initialise combination-test logging
+    from .logging_config import TestLogger
+
+    global _test_logger
+    log_dir = config.getoption("--combo-log-dir", default=None)
+    _test_logger = TestLogger(log_dir=log_dir)
+
```
Comment (tengqm) [inline tests/combination/conftest.py:82]: Well ... here is another logger, besides the standard one, why?

File: tests/combination/conftest.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_runtest_logreport(report):
+    """Log the outcome of each test (only for the 'call' phase)."""
+    if _test_logger is None:
+        return
+    if report.when == "call":
+        error_msg = None
+        if report.failed and report.longrepr:
```
Comment (tengqm) [inline tests/combination/conftest.py:121]: There are test cases that are "skipped" during the first few lines ..

File: tests/combination/conftest.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    x[inf_mask] = float("inf")
+    return x
+
+
+# ============================================================================
+# Helper to respect QUICK_MODE in tests
+# ============================================================================
+
+
+def skip_if_quick_mode(reason="Skipping in QUICK_MODE"):
```
Comment (tengqm) [inline tests/combination/conftest.py:309]:
```suggestion
def skip_if_quick_mode(reason="Skipped in QUICK_MODE"):
```

File: tests/combination/logging_config.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+# Setup / Teardown  (mirrors logging_utils.py patterns)
+# ---------------------------------------------------------------------------
+
+
+def _remove_file_handlers(logger):
+    """Remove and close FileHandlers owned by this module."""
+    for h in list(logger.handlers):
+        if isinstance(h, logging.FileHandler) and getattr(h, "_flaggems_owned", False):
+            h.close()
+            logger.removeHandler(h)
```
Comment (tengqm) [inline tests/combination/logging_config.py:69]: The existing logging facility is already a mess ... can we try making it worse?
